### PR TITLE
Improve handling of 8bit message content and `message/global` 

### PR DESCRIPTION
**Pull request by rsto**
This updates how Cyrus handles messages with 8bit content, both in headers and body parts values. This PR consists of a number of commits, some of which are refactoring only. The main functional changes are listed below:

- JMAP Email/set{create} now uses the same logic for serializing JMAP Blobs and JMAP EmailBodyValues to MIME. It updates how the content transfer encoding for MIME bodies is chosen and rewrites inline plain text if necessary.
- JMAP Email/set{create} and Email/{get} now both NFC-normalize internationalized email addresses.
- JMAP Email/{get} also NFC-normalizes `header:Xxx:asText` property values. JMAP Email/set{create} still q-encodes `header:Xxx:asText` property values but NFC-normalizes the value before encoding.
- JMAP Email/query and squatter now support internationalized email address domains queries in both Unicode and ASCII form.
- The minor mailbox cache now stores the content transfer encoding of body parts having type `message/rfc822` or `message/global`. The cache version is updated to `13`.

Of all changes, the ones done to the MIME serialization of JMAP Emails is the most critical: once the MIME message got generated its JMAP Email input as supplied by the client gets thrown away. Accordingly, this PR contains a bunch of test cases to trigger each special condition in the JMAP Email to MIME logic.

### CODE REVIEW

**Comment by ksmurchison** [comment]
@rsto I tried to give this a good review while on a train to Rome, but its been a long day and I couldn't focus enough to give it justice.  I probably can't get back to this until my flight home on June 8.

If you need it reviewed sooner, maybe someone else on the team can take a look.

**Review by elliefm — COMMENTED** [review]
Looks good, a few nits and questions though.

I'm not sure my review of the Xapian changes was quite comprehensive, I don't use C++ much, so I won't notice subtle mistakes in it.

File: cassandane/tiny-tests/JMAPEmail/email_set_encode_plain_text_attachment [hunk 1]
```diff
@@ -49,8 +49,32 @@ sub test_email_set_encode_plain_text_attachment
             fetchAllBodyValues => JSON::true,
         }, 'R2' ],
     ]);
-    my $subPart = $res->[1][1]{list}[0]{bodyStructure}{subParts}[1];
-    $self->assert_str_equals(' QUOTED-PRINTABLE', $subPart->{'header:Content-Transfer-Encoding'});
+
+    my ($maj, $min) = Cassandane::Instance->get_version();
+    if ($maj > 3 || ($maj == 3 && $min >= 9)) {
+        # Cyrus >= 3.9 sanitizes bare LF in inlined text bodies
+        xlog $self, "Assert that bare LF in inlined plain text gets expanded to CR LF";
```
[the remark this answers is not in the document]
Comment (rsto) [reply cassandane/tiny-tests/JMAPEmail/email_set_encode_plain_text_attachment:56]: Great thanks. Done.

File: imap/jmap_mail.c [hunk 2]
```diff
@@ -9920,7 +9861,13 @@ static struct emailpart *_emailpart_parse(jmap_req_t *req,
     }
 
     /* Finalize part definition */
-    part->jbody = json_incref(bodyValue);
+    if (bodyValue) {
+        json_t *jval = json_object_get(bodyValue, "value");
+        if (json_is_string(jval)) {
+            buf_init_ro(&part->bodyvalue, json_string_value(jval),
+                        json_string_length(jval));
```
[the remark this answers is not in the document]
Comment (rsto) [reply imap/jmap_mail.c:9881]: Great thanks. Done.

File: cassandane/tiny-tests/JMAPEmail/email_set_text_crlf [hunk 3]
```diff
@@ -30,5 +29,12 @@ sub test_email_set_text_crlf
...
     my $got = $ret->{bodyValues}{$ret->{textBody}[0]{partId}}{value};
-    $self->assert_str_equals($want, $got);
+
+    my ($maj, $min) = Cassandane::Instance->get_version();
+    if ($maj > 3 || ($maj == 3 && $min >= 9)) {
+        $self->assert_str_equals("ab\ncde\nfgh\nij", $got);
+    } else {
+        # pre-3.9 Cyrus JMAP discarded bare CR in text bodies
+        $self->assert_str_equals("ab\ncdefgh\nij", $got);
+    }
```
[the remark this answers is not in the document]
Comment (rsto) [reply cassandane/tiny-tests/JMAPEmail/email_set_text_crlf:39]: Done.

File: imap/jmap_mail.c [hunk 4]
```diff
@@ -10497,212 +10444,481 @@ static void _parse_email(jmap_req_t *req,
...
+        if (++line_len > MIME_MAX_LINE_LENGTH)
+            body_flags |= MIMEBODY_HAS_LONG_LINES;
+
+        if (c == '\r' || c == '\n') {
+            body_flags |= MIMEBODY_HAS_BARE_LF_CR;
+        }
+        else if (c == '\0') {
+            body_flags |= MIMEBODY_HAS_NUL;
+        }
+        else if (c && (c <= 0x8 || (c >= 0xb && c <= 0x1f) || c == 0x7f)) {
```
[the remark this answers is not in the document]
Comment (rsto) [reply imap/jmap_mail.c:10494]: Right, the outer check for `c` is unnecessary. I changed the code to check for the lower bound for `c <= 0x8`, it's not necessary as you remarked too, but I might make it less brittle if this code gets refactored (we don't want a zero to contribute as ASCII CTRL).

File: imap/message.c [hunk 5]
```diff
@@ -2518,8 +2523,9 @@ static void message_write_section(struct buf *buf, const struct body *body)
 {
     int part;
 
-    if (strcmp(body->type, "MESSAGE") == 0
-        && strcmp(body->subtype, "RFC822") == 0) {
+    if (!strcmp(body->type, "MESSAGE")
+        && (!strcmp(body->subtype, "RFC822") ||
+            !strcmp(body->subtype, "GLOBAL"))) {
```
[the remark this answers is not in the document]
Comment (rsto) [reply imap/message.c]: Thanks, done.

**Review by elliefm — APPROVED** [review]

**Comment by rsto** [comment]
This now also changes the Xapian backend to normalize any text using NFKC with CaseFolding before indexing and querying. It bumps the Xapian index version to 17. The default stemmer keeps using lowercase Cyrus Search Form.

If any of the existing indexes contains version 16 or less, then queries search for any terms that either match the NFKC-normalized term or the unnormalized term.

**Review by ksmurchison — APPROVED** [review]
LGTM

**Comment by rsto** [comment]
This now also includes https://github.com/cyrusimap/cyrus-imapd/pull/4977

**Review by brong — APPROVED** [review]
This is great work, thanks!  Is it at a good point for merging now?

**Comment by rsto** [comment]
Let's go for it. Merged.

**Pull request merged**